vochain/app.go

```diff
@@ -474,25 +496,75 @@ func (*BaseApplication) Query(_ context.Context,
 // the ResponsePrepareProposal call. The logic modifying the raw proposal MAY be non-deterministic.
 func (app *BaseApplication) PrepareProposal(ctx context.Context,
 	req *abcitypes.RequestPrepareProposal) (*abcitypes.ResponsePrepareProposal, error) {
-	// TODO: Prepare Proposal should check the validity of the transactions for the next block.
-	// Currently they are executed by CheckTx, but it does not allow height to be passed in.
-	validTxs := [][]byte{}
+	app.prepareProposalLock.Lock()
+	defer app.prepareProposalLock.Unlock()
+	type txInfo struct {
+		Data      []byte
+		Addr      *ethcommon.Address
+		Nonce     uint32
+		DecodedTx *vochaintx.Tx
+	}
+
+	validTxInfos := []txInfo{}
 	for _, tx := range req.GetTxs() {
-		resp, err := app.CheckTx(ctx, &abcitypes.RequestCheckTx{
-			Tx: tx, Type: abcitypes.CheckTxType_New,
+		vtx := new(vochaintx.Tx)
+		if err := vtx.Unmarshal(tx, app.ChainID()); err != nil {
+			// invalid transaction
+			log.Warnw("could not unmarshal transaction", "err", err)
+			continue
+		}
+		senderAddr, nonce, err := app.TransactionHandler.ExtractNonceAndSender(vtx)
+		if err != nil {
+			log.Warnw("could not extract nonce and/or sender from transaction", "err", err)
+			continue
+		}
+
+		validTxInfos = append(validTxInfos, txInfo{
+			Data:      tx,
+			Addr:      senderAddr,
+			Nonce:     nonce,
+			DecodedTx: vtx,
 		})
-		if err != nil || resp.Code != 0 {
+	}
+
+	// Sort the transactions based on the sender's address and nonce
+	sort.Slice(validTxInfos, func(i, j int) bool {
+		if validTxInfos[i].Addr == nil && validTxInfos[j].Addr != nil {
```

Comment:
Before the sort is executed, all the transactions included in `validTxInfos` are going to have a non `nil` address. I think the nil checks on the sort can be avoided.